mrmustard/lab_dev/circuit_components.py

```diff
@@ -181,6 +191,48 @@ def __getitem__(self, idx: Union[Mode, Sequence[Mode]]):
         ret._parameter_set = self.parameter_set
         return ret
 
+    def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other``, without adding adjoints.
+        """
+        # initialized the ``Wires`` of the returned component
+        wires_ret = self.wires @ other.wires
+
+        # find the indices of the wires being contracted on the bra side
```

Comment:
@ziofil to get to this code, I copy/pasted your code and did two things:
- Added logic to find the indices of the contracted wires on bra and ket side separately. Doing it as in your PR (see the ``common`` array) gave a bug with, e.g., ``dgate @ attenuator``, where ``dgate`` has no bra wire
- There was as issue with how you were finding the indices to reorder the final representation, what used to be ``before_ids``. I fixed it
Feel free to ask questions if you find more differences

Reply:
[@ziofil]: "Why do you need to find the indices for ket and bra separately?"

On Filippo's "state_barebones" branch, try:
```
from mrmustard.lab_dev.transformations import *

dgate = Dgate(1, modes=[0])
attenuator = Attenuator(1, modes=[0])

dgate @ attenuator
```

The last line gives the following error: 
```
File ~/git/MrMustard/mrmustard/physics/bargmann.py:140, in complex_gaussian_integral(Abc, idx_z, idx_zconj, measure)
    138 A, b, c = Abc
    139 if len(idx_z) != len(idx_zconj):
--> 140     raise ValueError("idx_z and idx_zconj must have the same length")
    141 n = len(idx_z)
    142 idx = tuple(idx_z) + tuple(idx_zconj)

ValueError: idx_z and idx_zconj must have the same length
```

The issue is caused by the first 4 lines in matmul:
```
common = set(self.wires.output.modes).intersection(other.wires.input.modes)
new_wires = self.wires >> other.wires
idx_z = self.wires[common].output.indices
idx_zconj = other.wires[common].input.indices
``` 
Here, ``common=[0]``, ``idx_z=[0]`` and ``idx_zconj=[1,3]``. So it's trying to contract one wire (the ket output of the Dgate) with two wires (the ket in and bra in of the attenuator). 
Finding the indices separately for ket and bra overcomes the issue, because the code will recognize that there are no wires to be contracted on the bra side

Reply:
ah I see! Thanks for the explanation